File: android/app/src/androidTest/java/net/ambitious/android/mealrelay/FoodClassifierTest.java
package net.ambitious.android.mealrelay;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageDecoder;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public final class FoodClassifierTest {
  @Test
  public void foodFixtureIsAccepted() throws IOException {
    Context context = InstrumentationRegistry.getInstrumentation().getContext();
    try (
        InputStream inputStream = context.getAssets().open("food.jpg");
        FoodClassifier classifier = new FoodClassifier(
            InstrumentationRegistry.getInstrumentation().getTargetContext())
    ) {
      Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
      assertTrue(classifier.isFood(bitmap));
      bitmap.recycle();
    }
  }

  @Test
  public void nonFoodFixtureIsRejected() throws IOException {
    Context context = InstrumentationRegistry.getInstrumentation().getContext();
    try (
        InputStream inputStream = context.getAssets().open("non_food.jpg");
        FoodClassifier classifier = new FoodClassifier(
            InstrumentationRegistry.getInstrumentation().getTargetContext())
    ) {
      Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
      assertFalse(classifier.isFood(bitmap));
      bitmap.recycle();
    }
  }

  @Test
  public void hardwareBitmapIsAccepted() throws IOException {
    Context context = InstrumentationRegistry.getInstrumentation().getContext();
    try (
        InputStream inputStream = context.getAssets().open("food.jpg");
        FoodClassifier classifier = new FoodClassifier(
            InstrumentationRegistry.getInstrumentation().getTargetContext())
    ) {
      ImageDecoder.Source source = ImageDecoder.createSource(
          ByteBuffer.wrap(inputStream.readAllBytes()));
      Bitmap bitmap = ImageDecoder.decodeBitmap(
          source,
          (decoder, information, imageSource) -> {
            decoder.setAllocator(ImageDecoder.ALLOCATOR_HARDWARE);
            decoder.setTargetSize(224, 224);
          });
      assertTrue(bitmap.getConfig() == Bitmap.Config.HARDWARE);
      assertTrue(bitmap.getWidth() == 224 && bitmap.getHeight() == 224);
      assertTrue(classifier.isFood(bitmap));
      bitmap.recycle();
    }
  }
}

File: android/app/src/androidTest/java/net/ambitious/android/mealrelay/FoodClassifierTest.kt
package net.ambitious.android.mealrelay

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageDecoder
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import java.nio.ByteBuffer
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class FoodClassifierTest {
  @Test
  fun foodFixtureIsAccepted() {
    val instrumentation = InstrumentationRegistry.getInstrumentation()
    instrumentation.context.assets.open("food.jpg").use { inputStream ->
      FoodClassifier(instrumentation.targetContext).use { classifier ->
        val bitmap = requireNotNull(BitmapFactory.decodeStream(inputStream))
        assertTrue(classifier.isFood(bitmap))
        bitmap.recycle()
      }
    }
  }

  @Test
  fun nonFoodFixtureIsRejected() {
    val instrumentation = InstrumentationRegistry.getInstrumentation()
    instrumentation.context.assets.open("non_food.jpg").use { inputStream ->
      FoodClassifier(instrumentation.targetContext).use { classifier ->
        val bitmap = requireNotNull(BitmapFactory.decodeStream(inputStream))
        assertFalse(classifier.isFood(bitmap))
        bitmap.recycle()
      }
    }
  }

  @Test
  fun hardwareBitmapIsAccepted() {
    val instrumentation = InstrumentationRegistry.getInstrumentation()
    instrumentation.context.assets.open("food.jpg").use { inputStream ->
      FoodClassifier(instrumentation.targetContext).use { classifier ->
        val source = ImageDecoder.createSource(ByteBuffer.wrap(inputStream.readAllBytes()))
        val bitmap = ImageDecoder.decodeBitmap(source) { decoder, _, _ ->
          decoder.allocator = ImageDecoder.ALLOCATOR_HARDWARE
          decoder.setTargetSize(224, 224)
        }
        assertEquals(Bitmap.Config.HARDWARE, bitmap.config)
        assertEquals(224, bitmap.width)
        assertEquals(224, bitmap.height)
        assertTrue(classifier.isFood(bitmap))
        bitmap.recycle()
      }
    }
  }
}
